lib/api/model/submessage.dart

```diff
@@ -313,3 +346,158 @@ class UnknownPollEventSubmessage extends PollEventSubmessage {
   @override
   Map<String, Object?> toJson() => json;
 }
+
+/// States of a poll Zulip widget.
+///
+/// See also:
+/// - https://zulip.com/help/create-a-poll
+/// - https://github.com/zulip/zulip/blob/304d948416465c1a085122af5d752f03d6797003/web/shared/src/poll_data.ts
+class Poll {
+  /// Construct a poll from submessages.
+  ///
+  /// For a poll Zulip widget, the first submessage's content contains a
+  /// [PollWidgetData], and all the following submessages' content each contains
+  /// a [PollEventSubmessage].
+  factory Poll.fromSubmessages({
+    required PollWidgetData widgetData,
+    required Iterable<Submessage> pollEventSubmessages,
+    required int messageSenderId,
+  }) {
+    final poll = Poll._(
+      messageSenderId: messageSenderId,
+      question: widgetData.extraData.question,
+      options: widgetData.extraData.options,
+    );
+
+    for (final submessage in pollEventSubmessages) {
+      final event = PollEventSubmessage.fromJson(jsonDecode(submessage.content) as Map<String, Object?>);
+      poll.applyEvent(submessage.senderId, event);
+    }
+    return poll;
+  }
+
+  Poll._({
+    required this.messageSenderId,
+    required this.question,
+    required List<String> options,
+  }) {
+    for (int index = 0; index < options.length; index += 1) {
+      // Initial poll options use a placeholder senderId.
+      // See [PollEventSubmessage.optionKey] for details.
+      _addOption(senderId: null, idx: index, option: options[index]);
+    }
+  }
+
+  final int messageSenderId;
+  String question;
+
+  /// The limit of options any single user can add to a poll.
+  ///
+  /// See https://github.com/zulip/zulip/blob/304d948416465c1a085122af5d752f03d6797003/web/shared/src/poll_data.ts#L69-L71
+  static const _maxIdx = 1000;
+
+  Iterable<PollOption> get options => _options.values;
+  /// Contains the text of all options from [_options].
+  final Set<String> _existingOptionTexts = {};
+  final Map<OptionKey, PollOption> _options = {};
+
+  void handleSubmessageEvent(SubmessageEvent event) {
+    final Object? decodedContent;
+    try {
+      decodedContent = jsonDecode(event.content);
+    } catch (e) {
+      assert(debugLog('Invalid JSON from submessage event data for poll: $e\n${jsonEncode(event)}')); // TODO(log)
+      return;
+    }
+
+    final PollEventSubmessage? pollEventSubmessage;
+    try {
+      pollEventSubmessage = PollEventSubmessage.fromJson(decodedContent as Map<String, Object?>);
+    } catch (e) {
+      assert(debugLog('Malformed submessage event data for poll: $e\n${jsonEncode(event)}')); // TODO(log)
+      return;
+    }
```

Comment:
Oh one more comment: we can simplify this a bit by making this just one try/catch that doesn't distinguish between whether it was the `jsonDecode` or the `fromJson` call that threw. Conceptually those are two phases of a single operation anyway: deserialize a string into a PollEventSubmessage.

Similarly if you walk through the logic in ApiConnection.send, I'm pretty sure the behavior there is that both kinds of failures throw very similar exceptions (instances of MalformedServerResponseException). So it's not worth getting more fine-grained here than we are there, the main place we deserialize JSON from the server.

(following up on https://github.com/zulip/zulip-flutter/pull/885#discussion_r1716404535 )